File: frontend/src/services/bookingService.js
import api from './api'

const bookingService = {

  // Create a new booking
  async createBooking(data) {
    const { data: res } = await api.post('/bookings', data)
    return res.data
  },

  // Get current user's bookings (paginated)
  async getMyBookings(page = 0, size = 10) {
    const { data: res } = await api.get('/bookings/my', { params: { page, size } })
    return res.data
  },

  // Admin: get all bookings
  async getAllBookings(status = null, page = 0, size = 20) {
    const params = { page, size }
    if (status) params.status = status
    const { data: res } = await api.get('/bookings/admin', { params })
    return res.data
  },

  // Get single booking
  async getBookingById(id) {
    const { data: res } = await api.get(`/bookings/${id}`)
    return res.data
  },

  // Approve a booking (admin/manager)
  async approveBooking(id, remarks = '') {
    const { data: res } = await api.put(`/bookings/${id}/approve`, { remarks })
    return res.data
  },

  // Reject a booking (admin/manager)
  async rejectBooking(id, reason) {
    const { data: res } = await api.put(`/bookings/${id}/reject`, { reason })
    return res.data
  },

  // Cancel a booking
  async cancelBooking(id) {
    const { data: res } = await api.put(`/bookings/${id}/cancel`)
    return res.data
  },

  // Permanently delete a PENDING booking (before admin review)
  async deleteBooking(id) {
    const { data: res } = await api.delete(`/bookings/${id}`)
    return res.data
  },

  // Get available time slots for a resource on a date
  async getAvailableSlots(resourceId, date) {
    const { data: res } = await api.get('/bookings/slots', {
      params: { resourceId, date }
    })
    return res.data  // array of { startTime, endTime, available }
  },

  // Check if a time slot has a conflict
  async checkConflict(resourceId, date, startTime, endTime) {
    const { data: res } = await api.get('/bookings/check-conflict', {
      params: { resourceId, date, startTime, endTime }
    })
    return res.data  // boolean
  },
}

export default bookingService

File: backend/src/main/java/com/smartcampus/modules/booking/service/BookingService.java
package com.smartcampus.modules.booking.service;

import com.smartcampus.modules.booking.dto.BookingDTO;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

import java.time.LocalDate;
import java.util.List;

public interface BookingService {

    BookingDTO.Response createBooking(BookingDTO.CreateRequest request, Long userId);

    BookingDTO.Response approveBooking(Long bookingId, Long adminId, String remarks);

    BookingDTO.Response rejectBooking(Long bookingId, Long adminId, String reason);

    BookingDTO.Response cancelBooking(Long bookingId, Long userId);

    void deleteBooking(Long bookingId, Long userId);

    BookingDTO.Response getBookingById(Long bookingId, Long userId);

    Page<BookingDTO.Response> getMyBookings(Long userId, Pageable pageable);

    Page<BookingDTO.Response> getAllBookings(String status, Pageable pageable);

    List<BookingDTO.TimeSlot> getAvailableTimeSlots(Long resourceId, LocalDate date);

    boolean checkConflict(Long resourceId, LocalDate date, java.time.LocalTime startTime, java.time.LocalTime endTime);
}
